### DISCUSSION TITLE
[v5] Missing relationship intellisense inside model methods

### DISCUSSION THREAD

**Discussion by rrmesquita** [adonisjs/core#1872] [Help] [answered]
What is the proper way to fetch all records for a given relationship from inside the model?

Even though it actually works as expected, the following code:

```ts
public async getRoles () {
  return await this.related('roles').query()
}
```

Will show the following intellisense error on `roles` string:

`Argument of type 'string' is not assignable to parameter of type 'ExtractModelRelations<this>'.ts(2345)`

![image](https://user-images.githubusercontent.com/30835404/98439086-f1ad8f00-20cd-11eb-9541-26e785793b23.png)

It's a bug or I'm missing something? 🤔 

UPDATE: Here's what expected when I hit ctrl + space:

![image](https://user-images.githubusercontent.com/30835404/98439133-80baa700-20ce-11eb-9e40-82e28b471a53.png)

**Comment by Reptiluka** [comment]
I don't know in which file you are (the model or a controller), but I think you can't simply use "this" as the wanted keyword User.
So I think the intellissense is right, use User instead of this 🙂

**Reply by rrmesquita** [reply]
Yeah, totally forgot to mention I'm in the User model class. But if you think that `this` represents the model instance (just like `new User()`), doesn't it make sense to grab its relationships as well?

**Reply by Reptiluka** [reply]
I think I already tried in the past to do something like you try, to "preload" for all model instance,  but finally I just use the model with the relationship tag, and in the controller do the others things.

If you found a solution write it here, it can be useful haha

**Comment by rrmesquita** [comment]
@thetutlage Any thoughts on that? I think the `ExtractModelRelations` type it's expecting an already instantiated model, maybe?

**Reply by thetutlage** [reply]
The issue is with Typescript. I raised it with them a while back, but its not fixed till now. https://github.com/microsoft/TypeScript/issues/37778

**Reply by eliferd** [reply]
Apparently, it is now finally fixed on TypeScript 5.1

**Comment by WeyIin** [comment]
Inside model:

```
  @belongsTo(() => Roles)
  public roles: BelongsTo<typeof Roles>

public getRoles() {
return this.roles
}

```

roles relationship has to be preloaded.

This is how I use it and unsure if it's right or wrong approach

**Comment by rrmesquita** [comment, marked as answer]
I've found a simple and clever workaround for this issue on [this comment](https://github.com/adonisjs/core/discussions/1555#discussioncomment-79801).

Basically, aliasing `this` does the trick.

```typescript
public async getRoles () {
  const user: User = this

  return await user.related('roles').query()
}
```

**Reply by motvicka** [reply]
Or just
```ts
public async getRoles () {
  return await (user as User).related('roles').query()
}
```

**Reply by rodrigorz** [reply]
I think i've found the correct way [here](https://www.typescriptlang.org/docs/handbook/2/classes.html#this-parameters).

`... we can add a this parameter to method definitions to statically enforce that the method is called correctly:`

```ts
class User {
  // ...

  public async getRoles (this: User, status: string) {
    return await this.related('roles').query().where('status', status).exec(); // now works
  }
}

const roles = await user.getRoles('active'); // this parameter will not be needed when call the method, just the other params
```

**Reply by rrmesquita** [reply]
"This" is a really cool feature from TypeScript, thank you for sharing @rodrigorz!

**Reply by inghggw** [reply]
thanks!!
